File: desktop/lib/models/quotation.dart
import 'entities.dart';

/// One line of an offer.
class QuotationLine {
  const QuotationLine({
    required this.id,
    required this.lineNumber,
    required this.productId,
    required this.description,
    required this.quantity,
    required this.unitPrice,
    required this.discountPercent,
    this.discountSource = '',
    this.freeQuantity = '0',
    required this.discountAmount,
    required this.taxAmount,
    required this.netAmount,
    required this.remarks,
  });

  final String id;
  final int lineNumber;
  final String productId;

  /// What the goods are called. Shown instead of the product id: a customer
  /// reading a quotation wants the item named, not a UUID.
  final String description;
  final String quantity;
  final String unitPrice;

  /// The rate that was quoted, which is what the editor puts back in the field.
  /// Only the resulting amount was parsed before, so revising a discounted
  /// line silently re-sent it at full price.
  final String discountPercent;

  /// `percent`/`amount` when typed, else where the server resolved it from;
  /// empty for a line stored before the source was recorded.
  final String discountSource;

  /// Thrown in with this line, charged for at nothing.
  final String freeQuantity;
  final String discountAmount;
  final String taxAmount;
  final String netAmount;
  final String remarks;

  factory QuotationLine.fromJson(Json json) => QuotationLine(
        id: stringValue(json['id']),
        lineNumber: (json['line_number'] as num?)?.toInt() ?? 0,
        productId: stringValue(json['product_id']),
        description: stringValue(json['description']),
        quantity: stringValue(json['quantity']),
        unitPrice: stringValue(json['unit_price']),
        discountPercent: stringValue(json['discount_percent']),
        discountSource: stringValue(json['discount_source']),
        freeQuantity: stringValue(json['free_quantity']).isEmpty
            ? '0'
            : stringValue(json['free_quantity']),
        discountAmount: stringValue(json['discount_amount']),
        taxAmount: stringValue(json['tax_amount']),
        netAmount: stringValue(json['net_amount']),
        remarks: stringValue(json['remarks']),
      );
}

/// A price offered to a customer before anything is sold.
///
/// A quotation commits nothing: no stock is reserved, no balance moves, no
/// journal is written. All of that happens at conversion, on the order it
/// becomes, which is why the screen is careful never to imply otherwise.
class Quotation {
  const Quotation({
    required this.id,
    this.version = 0,
    required this.customerId,
    required this.branchId,
    required this.warehouseId,
    required this.quotationNumber,
    required this.quotationDate,
    required this.validUntil,
    this.createdAt = '',
    required this.customerReference,
    required this.paymentTerms,
    required this.deliveryTerms,
    required this.status,
    this.billDiscountPercent = '0',
    this.billDiscountSource = '',
    this.freightAmount = '0',
    this.freightWaivedAmount = '0',
    required this.subtotal,
    required this.taxTotal,
    required this.grandTotal,
    required this.convertedSalesOrderId,
    required this.convertedSalesOrderNumber,
    required this.declineReason,
    required this.cancelReason,
    required this.remarks,
    required this.isExpired,
    required this.canConvert,
    required this.lines,
  });

  final String id;

  /// The optimistic-concurrency version this record was read at, sent back
  /// as `If-Match` on save so a concurrent edit is refused rather than
  /// silently overwritten. Zero means the server published none, and the
  /// save then carries no precondition.
  final int version;
  final String customerId;
  final String branchId;
  final String warehouseId;
  final String quotationNumber;
  final String quotationDate;

  /// The last day the quoted prices stand.
  final String validUntil;
  final String createdAt;
  final String customerReference;
  final String paymentTerms;
  final String deliveryTerms;
  final String status;
  /// What was taken off the whole offer, as a rate. Zero means none.
  final String billDiscountPercent;

  /// Where that came from: `typed`, `promotion` or `none`. Empty on an offer
  /// saved before the server recorded it, when it could only have been typed.
  final String billDiscountSource;

  /// What the customer is charged for delivery, after any free-shipping
  /// offer took its share off.
  final String freightAmount;

  /// What a free-shipping offer took off the delivery charge that was asked
  /// for. The charge asked is [freightAmount] plus this.
  final String freightWaivedAmount;

  final String subtotal;
  final String taxTotal;
  final String grandTotal;
  final String convertedSalesOrderId;
  final String convertedSalesOrderNumber;
  final String declineReason;
  final String cancelReason;
  final String remarks;

  /// Both answered by the server rather than worked out here, so the client
  /// cannot disagree with it about whether an offer still stands.
  final bool isExpired;
  final bool canConvert;

  final List<QuotationLine> lines;

  bool get isDraft => status == 'DRAFT';
  bool get isSent => status == 'SENT';
  bool get isAccepted => status == 'ACCEPTED';
  bool get isConverted => status == 'CONVERTED';
  bool get isDeclined => status == 'DECLINED';
  bool get isCancelled => status == 'CANCELLED';

  /// Whether the offer is still open — nobody has decided and it has not
  /// lapsed. This is what "how much business is on the table" counts.
  bool get isOpen => (isDraft || isSent) && !isExpired;

  factory Quotation.fromJson(Json json) => Quotation(
        id: stringValue(json['id']),
        version: (json['version'] as num?)?.toInt() ?? 0,
        customerId: stringValue(json['customer_id']),
        branchId: stringValue(json['branch_id']),
        warehouseId: stringValue(json['warehouse_id']),
        quotationNumber: stringValue(json['quotation_number']),
        quotationDate: stringValue(json['quotation_date']),
        validUntil: stringValue(json['valid_until']),
        createdAt: stringValue(json['created_at']),
        customerReference: stringValue(json['customer_reference']),
        paymentTerms: stringValue(json['payment_terms']),
        deliveryTerms: stringValue(json['delivery_terms']),
        status: stringValue(json['status']),
        billDiscountPercent:
            stringValue(json['bill_discount_percent']).isEmpty
                ? '0'
                : stringValue(json['bill_discount_percent']),
        billDiscountSource: stringValue(json['bill_discount_source']),
        freightAmount: stringValue(json['freight_amount']).isEmpty
            ? '0'
            : stringValue(json['freight_amount']),
        freightWaivedAmount: stringValue(json['freight_waived_amount']).isEmpty
            ? '0'
            : stringValue(json['freight_waived_amount']),
        subtotal: stringValue(json['subtotal']),
        taxTotal: stringValue(json['tax_total']),
        grandTotal: stringValue(json['grand_total']),
        convertedSalesOrderId: stringValue(json['converted_sales_order_id']),
        convertedSalesOrderNumber:
            stringValue(json['converted_sales_order_number']),
        declineReason: stringValue(json['decline_reason']),
        cancelReason: stringValue(json['cancel_reason']),
        remarks: stringValue(json['remarks']),
        isExpired: boolValue(json['is_expired']),
        canConvert: boolValue(json['can_convert']),
        lines: [
          for (final dynamic line in json['lines'] is List ? json['lines'] : const [])
            if (line is Map) QuotationLine.fromJson(Map<String, dynamic>.from(line)),
        ],
      );
}

/// What a conversion produced: the quotation, and the order it became.
class QuotationConversion {
  const QuotationConversion({
    required this.quotation,
    required this.orderNumber,
  });

  final Quotation quotation;
  final String orderNumber;

  factory QuotationConversion.fromJson(Json json) {
    final dynamic order = json['order'];
    final dynamic data = json['data'];
    return QuotationConversion(
      quotation: Quotation.fromJson(
        data is Map ? Map<String, dynamic>.from(data) : const {},
      ),
      orderNumber: order is Map
          ? stringValue(Map<String, dynamic>.from(order)['order_number'])
          : '',
    );
  }
}

/// One line's companions in a priced preview: what this customer last paid
/// for the product, and the stock free to promise where the offer ships from.
class QuotationPreviewLine {
  const QuotationPreviewLine({
    required this.lineNumber,
    required this.productId,
    required this.lastPrice,
    required this.lastInvoiceNumber,
    required this.lastInvoiceDate,
    required this.availableQuantity,
  });

  final int lineNumber;
  final String productId;

  /// Empty when the customer has never been billed for the product.
  final String lastPrice;
  final String lastInvoiceNumber;
  final String lastInvoiceDate;
  final String availableQuantity;

  factory QuotationPreviewLine.fromJson(Map<String, dynamic> json) =>
      QuotationPreviewLine(
        lineNumber: (json['line_number'] as num?)?.toInt() ?? 0,
        productId: '${json['product_id'] ?? ''}',
        lastPrice: json['last_price'] == null ? '' : '${json['last_price']}',
        lastInvoiceNumber: '${json['last_invoice_number'] ?? ''}',
        lastInvoiceDate: '${json['last_invoice_date'] ?? ''}',
        availableQuantity: '${json['available_quantity'] ?? '0'}',
      );
}

/// An offer priced exactly as saving it would, from
/// `POST /api/v1/quotations/preview`: the draft the save would store (its
/// rates, discounts, tax and number), how its tax splits, and each line's
/// companions. Nothing is saved.
class QuotationPreviewRecord {
  const QuotationPreviewRecord({
    required this.quotation,
    required this.interstate,
    required this.lines,
  });

  final Quotation quotation;

  /// IGST across states; CGST and SGST within one.
  final bool interstate;
  final List<QuotationPreviewLine> lines;

  factory QuotationPreviewRecord.fromJson(Map<String, dynamic> json) =>
      QuotationPreviewRecord(
        quotation: Quotation.fromJson(
          Map<String, dynamic>.from(json['quotation'] as Map? ?? const {}),
        ),
        interstate: json['interstate'] == true,
        lines: [
          for (final dynamic line in json['lines'] as List? ?? const [])
            if (line is Map)
              QuotationPreviewLine.fromJson(Map<String, dynamic>.from(line)),
        ],
      );
}

File: backend/app/quotation/models/quotation.py
"""Sales quotation persistence models.

A quotation is an offer, and an offer commits nothing. It reserves no stock,
puts nothing on a customer's account and writes no journal -- which is the
whole difference between it and the sales order it may become. Everything the
firm actually promises happens at conversion, through
``SalesOrderService.create_order``, so the rules that govern an order are
applied when the order exists rather than months earlier when somebody quoted
a price.

The one thing a quotation owns that an order does not is an expiry: a price
offered in April is not a price offered in December, and ``valid_until`` is
what stops a stale quote being turned into an order at last year's rate.
"""

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from sqlalchemy import (
    Date,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database.entity import BaseEntity
from app.core.database.types import UUIDType


class SalesQuotation(BaseEntity):
    """Store one quotation header."""

    __tablename__ = "sales_quotations"
    __table_args__ = (
        UniqueConstraint(
            "firm_id",
            "quotation_number",
            name="UQ_sales_quotations_firm_quotation_number",
        ),
        Index("IX_sales_quotations_firm_status", "firm_id", "status"),
        Index("IX_sales_quotations_firm_date", "firm_id", "quotation_date"),
        Index("IX_sales_quotations_firm_customer", "firm_id", "customer_id"),
        Index("IX_sales_quotations_firm_valid_until", "firm_id", "valid_until"),
    )

    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    customer_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("customers.id", ondelete="RESTRICT"), nullable=False
    )
    salesman_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("users.id", ondelete="RESTRICT")
    )
    territory_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("sales_territories.id", ondelete="RESTRICT")
    )
    branch_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("branches.id", ondelete="RESTRICT"), nullable=False
    )
    #: Carried so the order this becomes has one. A quotation reserves nothing
    #: from it -- naming the warehouse is a statement about where the goods
    #: would ship from, not a claim on them.
    warehouse_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("warehouses.id", ondelete="RESTRICT"), nullable=False
    )
    business_profile_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("business_profiles.id", ondelete="RESTRICT")
    )
    quotation_number: Mapped[str] = mapped_column(String(60), nullable=False)
    quotation_date: Mapped[date] = mapped_column(Date, nullable=False)
    #: The last day the quoted prices stand. Past it a quotation can be read
    #: and reissued but not turned into an order.
    valid_until: Mapped[date] = mapped_column(Date, nullable=False)
    customer_reference: Mapped[str | None] = mapped_column(String(80))
    reference_number: Mapped[str | None] = mapped_column(String(80))
    payment_terms: Mapped[str | None] = mapped_column(String(200))
    delivery_terms: Mapped[str | None] = mapped_column(String(200))
    currency_code: Mapped[str | None] = mapped_column(String(10))
    exchange_rate: Mapped[Decimal | None] = mapped_column(Numeric(18, 6))
    remarks: Mapped[str | None] = mapped_column(Text)
    #: The customer's standing discount when this document was raised. The
    #: rate is a starting point and every line may override it, so this says
    #: what it would have been rather than what any line was charged.
    customer_discount_percent: Mapped[Decimal] = mapped_column(
        Numeric(9, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    status: Mapped[str] = mapped_column(
        String(30), nullable=False, default="DRAFT", server_default="DRAFT"
    )
    #: A discount on the whole document, negotiated once rather than typed on
    #: every line. It comes off what the lines discounted to, and each line
    #: carries its share so the tax is charged on the discounted value.
    bill_discount_percent: Mapped[Decimal] = mapped_column(
        Numeric(9, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    bill_discount_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    #: What the customer is charged for getting the goods to them.
    #:
    #: **Part of the taxable value, not an extra on the end.** Delivery charged
    #: by the seller is ancillary to the supply of the goods, so it is taxed at
    #: the goods' own rate -- which is what apportioning it across the lines
    #: achieves. `additional_charges` sits outside the tax and stays that way:
    #: it is for genuinely non-taxable additions, and silently re-taxing it
    #: would change the meaning of every document that carries one.
    #:
    #: The mirror image of `bill_discount_amount`, and it uses the same
    #: `apportion`: one reduces each line's taxable value and the other raises
    #: it, and both give the rounding residual to the largest line so the
    #: shares sum exactly to the header figure.
    freight_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    line_discount_total: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    subtotal: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    tax_total: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    additional_charges: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    round_off: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    grand_total: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    sent_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    decided_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    converted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    #: The order this quotation became. A bare id with no foreign key, the
    #: convention every cross-document reference here follows.
    converted_sales_order_id: Mapped[UUID | None] = mapped_column(UUIDType())
    converted_sales_order_number: Mapped[str | None] = mapped_column(String(60))
    #: Why the customer said no. Kept because the answer to "why are we losing
    #: quotes" is not in any total.
    decline_reason: Mapped[str | None] = mapped_column(Text)
    cancel_reason: Mapped[str | None] = mapped_column(Text)


class SalesQuotationLine(BaseEntity):
    """Store one quotation line."""

    __tablename__ = "sales_quotation_lines"
    __table_args__ = (
        UniqueConstraint(
            "sales_quotation_id",
            "line_number",
            name="UQ_sales_quotation_lines_quotation_line",
        ),
        Index("IX_sales_quotation_lines_firm_product", "firm_id", "product_id"),
    )

    sales_quotation_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("sales_quotations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    line_number: Mapped[int] = mapped_column(Integer, nullable=False)
    product_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("products.id", ondelete="RESTRICT"), nullable=False
    )
    description: Mapped[str | None] = mapped_column(String(500))
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False)
    free_quantity: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    sales_uom_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("uoms.id", ondelete="RESTRICT")
    )
    inventory_uom_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("uoms.id", ondelete="RESTRICT")
    )
    packaging_type_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("packaging_types.id", ondelete="RESTRICT")
    )
    unit_price: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    discount_percent: Mapped[Decimal] = mapped_column(
        Numeric(9, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    discount_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    gross_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    tax_profile_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("tax_profiles.id", ondelete="RESTRICT")
    )
    tax_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    #: This line's share of the document's bill discount. Stored rather than
    #: derived at print time, because it is what the tax was computed on.
    bill_discount_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    #: This line's share of the document's freight, in proportion to what the
    #: line is worth after its own discounts. Stored rather than derived at
    #: print time, for the reason the bill discount's share is: the tax was
    #: charged on it, so the number has to survive.
    freight_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    net_amount: Mapped[Decimal] = mapped_column(
        Numeric(18, 4), nullable=False, default=Decimal("0"), server_default="0"
    )
    warehouse_id: Mapped[UUID | None] = mapped_column(
        UUIDType(), ForeignKey("warehouses.id", ondelete="RESTRICT")
    )
    remarks: Mapped[str | None] = mapped_column(Text)


class SalesQuotationAttachment(BaseEntity):
    """Store quotation attachments."""

    __tablename__ = "sales_quotation_attachments"

    sales_quotation_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("sales_quotations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    file_name: Mapped[str] = mapped_column(String(260), nullable=False)
    mime_type: Mapped[str | None] = mapped_column(String(120))
    file_path: Mapped[str] = mapped_column(String(1024), nullable=False)
    attachment_kind: Mapped[str] = mapped_column(
        String(40),
        nullable=False,
        default="QUOTATION_FILE",
        server_default="QUOTATION_FILE",
    )


class SalesQuotationNote(BaseEntity):
    """Store quotation notes."""

    __tablename__ = "sales_quotation_notes"

    sales_quotation_id: Mapped[UUID] = mapped_column(
        UUIDType(),
        ForeignKey("sales_quotations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    firm_id: Mapped[UUID] = mapped_column(
        UUIDType(), ForeignKey("firms.id"), nullable=False, index=True
    )
    note_type: Mapped[str] = mapped_column(
        String(30), nullable=False, default="INTERNAL", server_default="INTERNAL"
    )
    note: Mapped[str] = mapped_column(Text, nullable=False)
